x/attribute/keeper/keeper.go

```diff
@@ -435,3 +510,59 @@ func (k Keeper) PopulateAddressAttributeNameTable(ctx sdk.Context) {
 		k.IncAttrNameAddressLookup(ctx, attr.Name, attr.GetAddressBytes())
 	}
 }
+
+// DeleteExpiredAttributes find and delete expired attributes returns the total deleted
+// limit sets the max amount to delete in a call, 0 for not limit
+func (k Keeper) DeleteExpiredAttributes(ctx sdk.Context, limit int) int {
+	store := ctx.KVStore(k.storeKey)
+
+	iterator := store.Iterator(types.AttributeExpirationKeyPrefix, types.GetAttributeExpireTimePrefix(ctx.BlockTime()))
+	defer iterator.Close()
+	count := 0
+	for ; iterator.Valid(); iterator.Next() {
+		attrKey := types.GetAddrAttributeKeyFromExpireKey(iterator.Key())
+		bz := store.Get(attrKey)
+		if bz != nil {
+			var attribute types.Attribute
+			if err := k.cdc.Unmarshal(bz, &attribute); err == nil {
+				// delete attribute from store
+				store.Delete(attrKey)
+				// dec name to address lookup table count
+				k.DecAttrNameAddressLookup(ctx, attribute.Name, attribute.GetAddressBytes())
+
+				deleteExpirationEvent := types.NewEventAttributeExpiredDelete(attribute)
+				if err = ctx.EventManager().EmitTypedEvent(deleteExpirationEvent); err != nil {
+					ctx.Logger().Error(fmt.Sprintf("failed to emit typed event %v", err))
+				}
+				count++
+			} else {
+				ctx.Logger().Error(fmt.Sprintf("unable to unmarshal attribute to delete key: %v error: %v", attrKey, err))
+				store.Delete(iterator.Key())
+				continue
+			}
+		}
+
+		// delete the expiration lookup key
+		store.Delete(iterator.Key())
+		if limit != 0 && count >= limit {
+			break
+		}
+	}
+	return count
+}
+
+// AddAttributeExpireLookup safely adds attribute expire key to store, if expire date exists, else no-op
+func (k Keeper) AddAttributeExpireLookup(ctx sdk.Context, attr types.Attribute) {
+	expireKey := types.AttributeExpireKey(attr)
+	if expireKey != nil {
+		ctx.KVStore(k.storeKey).Set(expireKey, []byte{})
```

Comment:
Ok, I will make it private and pass in the store.  Also, the store does allow empty just not nil values for `value`, the key cannot be nil.
These are the two validation methods called at the Set https://github.com/provenance-io/cosmos-sdk/blob/3c797794e2c8edd432dbb1061de6c417cce4fd50/store/types/validity.go#L11

Set: https://github.com/provenance-io/cosmos-sdk/blob/3c797794e2c8edd432dbb1061de6c417cce4fd50/store/gaskv/store.go#L47

I believe we should keep it as an empty `value` which will reduce the gas cost calculated later in the `Set` method.

Reply:
Hmmm. I could have sworn `Set` would panic if provided an empty `value`. That's why we use `[]byte{0x01}` all over the place for the index entries.

That being said, The SDK's `store/types.AssertValidValue` function only errors when the value is `nil` (but an empty slice would be okay). And it looks like that's used in most places, at least on the SDK's level. The db adapters I looked at only checked for `nil` too. So maybe it's okay as empty.

Reply:
Yeah, before I made these methods and looked at their code.  I did think that it didn't allow empty arrays because we use []byte{0x01} all over the place.  However, it does make sense that you should be able to have an empty array in there.